### ISSUE TITLE
Wrong colours on yabasanshiro on RPI4 + hack to fix it 

### ISSUE THREAD

**Issue by joolswills** [devmiyax/yabause#737]
Just wanted to leave this here - as this is the upstream of the libretro core

https://github.com/libretro/yabause/issues/208

As you may be aware, the colours are wrong on the RPI4 with yabasanshiro.

It seems to need the calcs for the texelFetch calls to be rounded up to work - but I'm not sure the cause, or if it's an issue with the VC6 MESA shader compilation code, or a bug that only manifests on this platform.

My knowledge of this is not great.

Here's a patch to fix it - note that this may not be complete, and I do consider it a hack than a fix. But perhaps it is useful.

https://github.com/RetroPie/RetroPie-Setup/blob/5e97cc038e40661d7f97f9d5a0955ba0773b0da5/scriptmodules/libretrocores/lr-yabasanshiro/01_shader_hack_rpi4.diff

I wondered if it might be related to something like https://stackoverflow.com/questions/40574677/how-to-normalize-image-coordinates-for-texture-space-in-opengl

The shader code is quite hard to follow - conversions from int to float, then back to int - I suspect a reworking of the code may show up the issue more clearly.

I originally added round() calls to fix it, but adding 0.5 is equivalent. And makes for less brackets in the changes. There may be cases I've missed where there are still issues too in the source - but Im posting this here in the hope it will aid someone debugging the issue properly.

**Comment by devmiyax** [comment]
Thank you for you investigation and patch. Do you have screenshots how this patch improve color issues?

**Comment by pjft** [comment]
I can provide them tomorrow if needed, but basically without the patch the colors are very much in the realm of very dark hues or even black. Similar (or likely the same as) #704, though the description there only mentions "inversion" and it isn't really that. I don't have the games from those screenshots, so I can't compare.

With these adjustments the color are very much close to (if not the same as) what they should be in the RPi4 - so, if I were none the wiser, I'd say "it fixes it". I don't really have a frame of reference for how the real colors should be, other than memory of these games, but they all look as they should. I don't have the emulator running on any other platform to compare, though.

**Comment by pjft** [comment]
As I can't find an easy way to upload pictures here, this will have to do:

https://photos.google.com/share/AF1QipMX0GVretGMI8F1RA8Do9fFtKxbyMjV6QGtKJHt8EFOyyUrqHG7h3tcWU2IaieM1Q?key=OTVQMGwwYkUzdmhuR3ZWQ3VhWVVKcWhQWktJRk1R

I captured 5 different occasions in Die Hard Arcade, for instance, both without the patch and with this patch. They are sorted, hopefully, so that they're the following:

- Fox logo no-patch / patch
- Attract mode no-patch / patch
- Menu no-patch / patch
- In-game Intro no-patch / patch
- In-game gameplay no-patch / patch

Hope these help.

**Comment by joolswills** [comment]
Thanks for doing that.

**Comment by devmiyax** [comment]
![IMG_20200728_225010](https://user-images.githubusercontent.com/1174152/88676072-9e1ec300-d126-11ea-999e-445cc802022f.jpg)
Thank you. it works.  I pushed code for the standalone version to pi4 branch. https://github.com/devmiyax/yabause/tree/pi4
